Add examples to documentation for DataFrame.swaplevel

thordisstella opened the pull request:
- [x ] closes https://github.com/MarcoGorelli/pyladies-sprint-March-2021/issues/13
- [ ] tests added / passed
- [ ] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [ ] whatsnew entry

thordisstella committed:
Add examples to documentation for DataFrame.swaplevel

pep8speaks commented:
Hello @thordisstella! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pandas/core/frame.py`](https://github.com/pandas-dev/pandas/blob/7daa05c368dc6b20f12bfeeab5fea519e014e195/pandas/core/frame.py):

> [Line 6285:89](https://github.com/pandas-dev/pandas/blob/7daa05c368dc6b20f12bfeeab5fea519e014e195/pandas/core/frame.py#L6285): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (94 > 88 characters)
> [Line 6296:89](https://github.com/pandas-dev/pandas/blob/7daa05c368dc6b20f12bfeeab5fea519e014e195/pandas/core/frame.py#L6296): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (106 > 88 characters)
> [Line 6298:89](https://github.com/pandas-dev/pandas/blob/7daa05c368dc6b20f12bfeeab5fea519e014e195/pandas/core/frame.py#L6298): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (96 > 88 characters)
> [Line 6306:89](https://github.com/pandas-dev/pandas/blob/7daa05c368dc6b20f12bfeeab5fea519e014e195/pandas/core/frame.py#L6306): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (89 > 88 characters)

MarcoGorelli commented on pandas/core/frame.py:
```diff
@@ -6278,6 +6278,54 @@ def swaplevel(self, i: Axis = -2, j: Axis = -1, axis: Axis = 0) -> DataFrame:
+
+        Examples
+        --------
+        >>> df = pd.DataFrame({'Grade':['A','B','A','C']},
+        ...                   index=[['Final exam', 'Final exam', 'Coursework',
+        ...                              'Coursework'],
+        ...                         ['History','Geography','History','Geography'],
+        ...                         ['January','February','March','April']])
```
Could you reformat this according to the `black` formatter (to be consistent with most of the rest of the docs...we should probably run `blackdoc`, but that's for a separate issue/PR)

MarcoGorelli commented on pandas/core/frame.py:
```diff
@@ -6278,6 +6278,54 @@ def swaplevel(self, i: Axis = -2, j: Axis = -1, axis: Axis = 0) -> DataFrame:
+        We can also define explicitly which indices we want to swap by supplying values
+        for both i and j. Here, we for example swap the first and second indices.
+
+        >>> df.swaplevel(0,1)
```
can you add a space between the arguments? (`swaplevel(0, 1)`)

MarcoGorelli reviewed (changes requested):
Really nice example @thordisstella , thanks, have just left some minor comments

thordisstella commented:
> Really nice example @thordisstella , thanks, have just left some minor comments

Thanks for the review @MarcoGorelli! I addressed the comments, but when I ran `black` on the changes I'd made, nothing further was updated (after I'd addressed your other two comments). Please let me know if I've missed something here

MarcoGorelli commented:
Hey @thordisstella , thanks for updating! Sorry, I was unclear - `black` indeed won't modify what you've written, because it doesn't format code inside docstrings (barring some very minor things). I was suggesting to reformat the docstring the way that `black` would - one way to do this would be to paste the docstring into a temporary file (say, `t.py`):
```python
# t.py

"""
        >>> df = pd.DataFrame({'Grade':['A','B','A','C']},
        ...                   index=[['Final exam', 'Final exam', 'Coursework',
        ...                              'Coursework'],
        ...                         ['History','Geography','History','Geography'],
        ...                         ['January','February','March','April']])
"""
```
, running
```python
pip install blackdoc
blackdoc t.py --line-length 80
```
and then pasting the reformatted version back into this docstring of `swaplevel`

Finally, you should run
```console
$ python scripts/validate_docstrings.py pandas.DataFrame.swaplevel
```
and check it doesn't throw any extra errors compared to when run on `master` (in this case you prob. just need to reformat the code in the example and remove the extra blankline at the end of the docstring)

thordisstella commented:
Thanks, and belated happy IWD to you too! 😊 

Thanks for the clarification - I've updated the PR as you suggested so hopefully we should be good now 🤞

MarcoGorelli commented on pandas/core/frame.py:
```diff
@@ -6278,6 +6278,58 @@ def swaplevel(self, i: Axis = -2, j: Axis = -1, axis: Axis = 0) -> DataFrame:
+        History     Final exam  January         A
+        Geography   Final exam  February        B
+        History     Coursework  March           A
+        Geography   Coursework  April           C
+
         """
```
```suggestion
        Geography   Coursework  April           C
        """
```

MarcoGorelli reviewed (changes requested):
Thanks - there's just a blank line which needs removing (else it'll fail `scripts/validate_docstrings.py`), rest looks good

MarcoGorelli commented:
Hi @thordisstella - looks good, CI failures should be unrelated, could you fetch and merge upstream/master please?

MarcoGorelli committed:
Merge remote-tracking branch 'upstream/master' into pr/thordisstella/40269

MarcoGorelli reviewed (approved):
Pending green, looks good to me!

(a follow-up PR to add a similar example to `Series.swaplevel` and/or make a shared docstring would be welcome if you want)

thordisstella commented:
Thanks for the review, all the advice and for merging this @MarcoGorelli ! And apologies for the delay in reacting to your last comment. I'll take a look at the follow up PR you suggested either later this week or next week :)
